Readme update (#396)
coveo/coveo.analytics.js: README.md

## Loading and initializing the library in the browser

In order to ensure the tracking code is available on your webpage, the following code snippet needs to be added to the top of each page on which analytics are required. This will load the latest major version of coveo.analytics.js from a Coveo CDN. As of writing, the current major version is 2.

```html
<script>
    (function (c, o, v, e, O, u, a) {
        a = 'coveoua';
        c[a] =
            c[a] ||
            function () {
                (c[a].q = c[a].q || []).push(arguments);
            };
        c[a].t = Date.now();
        u = o.createElement(v);
        u.async = 1;
        u.src = e;
        O = o.getElementsByTagName(v)[0];
        O.parentNode.insertBefore(u, O);
    })(window, document, 'script', 'https://static.cloud.coveo.com/coveo.analytics.js/2/coveoua.js');
</script>

> coveoua('init', <COVEO_API_KEY>); // Replace <COVEO_API_KEY> with your api key</COVEO_API_KEY></COVEO_API_KEY>

Author: Thanks vscode! You're so helpful!

```

> Since calls to the coveo analytics service need to be authenticated, the library needs to be initialized with a Coveo api key which has push access to the [Analytics Data Domain](https://docs.coveo.com/en/1707/cloud-v2-administrators/privilege-reference#analytics-data-domain). You can create an API key from the [administration console](https://platform.cloud.coveo.com/admin/#/organization/api-access/) selecting the **Push** option box for the **Analytics Data** domain (see [Adding and Managing API Keys](https://docs.coveo.com/en/1718/cloud-v2-administrators/adding-and-managing-api-keys)).

Reviewer: Nice detail here 💯

    -   COVEO_API_KEY (mandatory): A valid api key.
    -   ENDPOINT (optional): A string specifying the desired analytics endpoint. The default value is https://analytics.cloud.coveo.com/rest/ua. In case your organization is HIPAA enabled, you should override with https://analyticshipaa.cloud.coveo.com/rest/ua.
-   `coveoua('init', <COVEO_API_KEY>, {endpoint: <ENDPOINT>, plugins: <PLUGINS>})`: Initializes the library with the given api key, endpoint and plugins. The following parameters are accepted
    -   COVEO_API_KEY (mandatory): A valid api key.
    -   ENDPOINT (optional): An object string specifying the desired analytics endpoint. The default value is https://analytics.cloud.coveo.com/rest/ua. In case your organization is HIPAA enabled, you should override with https://analyticshipaa.cloud.coveo.com/rest/ua.
    -   PLUGINS (optional): An array of known plugin names. See [plugins](#plugins) for more information.
-   `coveoua('set', <NAME>, <VALUE>)`: Attempts to inject an attribute with given name and value on every logged event. Some payloads may reject attributes they do not support.

> -   `coveoua('set', <OBJECT>)`: Attempts to inject all attributes and values of the given object on every logged event. Some payloads may reject attributes they do not support.

Reviewer: Would specifying an object override any previously specified name/value pairs? Perhaps worth mentioning if so,

```js
coveoua('send', 'click', {...});
```

You should be able to observe the click event being transmitted to the Coveo backend at `https://analytics.cloud.coveo.com/rest/ua/click` in the Developer tool's **Network** tab of your browser of choice.

## Sending commerce specific events

Commerce specific events such as product selections, shopping cart modifications and transactions are sent to Coveo in the compact [collect protocol](https://docs.coveo.com/en/l41i0031/build-a-search-ui/log-collect-events). Rather than explicitly assembling these payloads by hand, the eCommerce plugin provides compact apis to assemble and transmit the payloads. There are two event names that are specific to the eCommerce plugin:

-   `event`: A generic event, which has been assembled through different plugin actions.
-   `pageview`: An ecommerce specific pageview event which is automatically populated.

See the [Send an Event](https://docs.coveo.com/en/l3am0254/coveo-for-commerce/send-an-event) page for more information on the expected payloads for both of these.

> The full list of supported eCommerce plugin actions are listed in the [**Possible Actions** section of the Tracking Commerce Events page](https://docs.coveo.com/en/3188/coveo-for-commerce/commerce-data-health-implementation-guide#possible-actions). This page also provides information on how to log

Reviewer 2: ```suggestion
All supported actions are listed in the [**Product action type reference**](https://docs.coveo.com/en/l29e0540/coveo-for-commerce/commerce-events-reference#product-action-type-reference). This page also provides information on how to log
```

Author: This is a different list. eCommerce plugin actions are things like `ec:AddProduct` and `ec:addImpression`. I'm not sure if there a proper reference to all possible commands in the public doc. I can try to add them here.

Reviewer 2: You're right, we don't currently have an article that lists them all, they are however scattered throughout the documentation in the respective areas of usage.

## Running tests

1. From the command line through `npm run test`.
2. Debugging through VSCode debugger with the `Debug: Start Debugging` command, using the `Jest All` configuration.

## Storage and persistence

Coveo.analytics.js tracks interactions from the same browserclient, through a clientside provided uuid called a `clientId`. This clientId is initialized on first use and there are multiple options for persisting it's value:

-   Cookie storage, which supports top level domain storage. This means that the clientId for a.foo.com will be identical to the one on b.foo.com.
-   Local storage, which allows to store much more information client side, but has the drawback of not being able to access data across multiple top level domains.
-   Session storage, which has roughly the same limitation and capability as Local storage, except that it is cleared when the web browser tab is closed.

By default, coveoua will use both local storage and cookie storage to persist it's clientId. If your environment does not support local persistence, it's possible to write your own storage abstraction.

## Using coveo.analytics.js with React Native

> Since React Native does not run inside a browser, it cannot use cookies or the local/session storage that modern browsers provide. You must provide your own Storage implementation. Thankfully, there exist multiple packages to store data:

Reviewer: If using the library with react native becomes common, maybe an alternative could be to export the implementation for the developer to pass as a plugin?
